mixer/cmd/reset.go

```diff
@@ -0,0 +1,200 @@
+	if format, err = ioutil.ReadFile(filename); err != nil {
+		fail(err)
+	} else {
+		if strings.TrimSpace(string(format)) != b.State.Mix.Format {
+			b.State.Mix.Format = strings.TrimSpace(string(format))
+		}
+	}
+
+	// Change upstreamVersion file content to the previous version
+	var lastStableMixUpstreamVersion []byte
+	filename = filepath.Join(b.Config.Builder.ServerStateDir, "www", b.State.Mix.PreviousMixVer, "upstreamver")
+	if lastStableMixUpstreamVersion, err = ioutil.ReadFile(filename); err != nil {
+		fail(err)
+	} else {
+		if strings.TrimSpace(string(lastStableMixUpstreamVersion)) != b.UpstreamVer {
+			// Set the upstream version to the previous format's latest version
+			b.UpstreamVer = strings.TrimSpace(string(lastStableMixUpstreamVersion))
+			b.UpstreamVerUint32, err = parseUint32(b.UpstreamVer)
+			if err != nil {
+				fail(errors.Wrapf(err, "Couldn't parse upstream version"))
+			}
+			vFile := filepath.Join(b.Config.Builder.VersionPath, b.UpstreamVerFile)
+			if err = ioutil.WriteFile(vFile, []byte(b.UpstreamVer), 0644); err != nil {
+				fail(err)
+			}
+		}
+	}
+
+	// Change mixVersion to point to the last good build
+	if err := ioutil.WriteFile(filepath.Join(b.Config.Builder.VersionPath, b.MixVerFile), []byte(b.State.Mix.PreviousMixVer), 0644); err != nil {
+		fail(err)
+	}
+	b.MixVer = b.State.Mix.PreviousMixVer
+	b.MixVerUint32, err = parseUint32(b.MixVer)
+	if err != nil {
+		fail(errors.Wrapf(err, "Couldn't parse mixer version"))
+	}
+
+	// Make sure update/image/LAST_VER points to the last good build
+	if err = ioutil.WriteFile(filepath.Join(b.Config.Builder.ServerStateDir, "image", "LAST_VER"), []byte(b.State.Mix.PreviousMixVer), 0644); err != nil {
+		failf("Couldn't update LAST_VER file: %s", err)
+	}
+
+	// update the state.mix file
+	err = b.State.Save()
+	if err != nil {
+		failf("Couldn't update mixer.state file: %s", err)
+	}
+
+	// update www/version/latest_version file and check if sig file exists, if it does, regenerate the signature
+	// update www/version/formatXX/latest file and update the content, if signature exits , regenerate the signature
+	var skipSign bool
+	latestVerSigFilePath := filepath.Join(b.Config.Builder.ServerStateDir, "www", "version", "latest_version.sig")
+	if _, err = ioutil.ReadFile(latestVerSigFilePath); err != nil {
+		skipSign = true
+	}
+	err = b.UpdateLatest(skipSign)
+	if err != nil {
+		fail(err)
+	}
+
+	// if clean flag set
+	if resetCmdFlags.clean {
+		// Remove any folder inside update/image for versions above prevMixVer
+		files, err := ioutil.ReadDir(b.Config.Builder.ServerStateDir + "/image")
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		for _, f := range files {
+			dirNameInt, err := parseUint32(f.Name())
+			if err != nil {
+				continue
+			}
+			if dirNameInt > stateMixPreviousMixVersionUint32 {
+				err := os.RemoveAll(b.Config.Builder.ServerStateDir + "/image/" + f.Name())
+				if err != nil {
+					continue
```

Comment:
we should evaluate what was the error here and at least print a warning. I assume it will be common to forget to use sudo to run this command which should result in a permission error. It would be better to inform the user if something goes wrong.